File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/6110a86f4218962475ccd7a3475ce2eb.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"time"
)

type pt struct {
	x int
	y int
}

type pair struct {
	a int
	b int
}

func main() {
	rand.Seed(time.Now().UnixNano())

	const INF = int(1e9)
	const INF64 = int64(1e18)
	const MOD = INF + 7
	const EPS = 1e-9
	const PI = math.Acos(-1.0)

	const N = 100*1000 + 13

	var n, m, k int
	var h int64
	var a [N]pair

	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	read := func() bool {
		scanner.Scan()
		h, _ = strconv.ParseInt(scanner.Text(), 10, 64)
		scanner.Scan()
		n, _ = strconv.Atoi(scanner.Text())
		scanner.Scan()
		m, _ = strconv.Atoi(scanner.Text())
		scanner.Scan()
		k, _ = strconv.Atoi(scanner.Text())

		for i := 0; i < n; i++ {
			scanner.Scan()
			a[i].a, _ = strconv.Atoi(scanner.Text())
			scanner.Scan()
			a[i].b, _ = strconv.Atoi(scanner.Text())
		}

		return true
	}

	used := make([]bool, N)
	mn := make([]int64, N)
	cur := make([]pt, 0)

	add := func(x int64) {
		q := make([]int, 0)
		for i := 0; i < k; i++ {
			q = append(q, i)
		}
		for len(q) > 0 {
			v := q[0]
			q = q[1:]
			if mn[v]+x < mn[(v+x)%k] {
				mn[(v+x)%k] = mn[v] + x
				q = append(q, (v+x)%k)
			}
		}
	}

	recalc := func() {
		cur = make([]pt, 0)
		for i := 0; i < n; i++ {
			if !used[i] && mn[a[i].a%k] <= int64(a[i].a) {
				cur = append(cur, pt{a[i].b, -i})
			}
		}
	}

	solve := func() {
		for i := 0; i < k; i++ {
			mn[i] = h + 1
		}
		mn[1] = 1
		for i := 0; i < len(used); i++ {
			used[i] = false
		}
		recalc()
		for i := 0; i < m; i++ {
			scanner.Scan()
			t, _ := strconv.Atoi(scanner.Text())
			if t == 1 {
				scanner.Scan()
				x, _ := strconv.ParseInt(scanner.Text(), 10, 64)
				add(x)
				recalc()
			} else if t == 2 {
				scanner.Scan()
				x, _ := strconv.Atoi(scanner.Text())
				scanner.Scan()
				y, _ := strconv.Atoi(scanner.Text())
				x--
				if mn[a[x].a%k] <= int64(a[x].a) {
					for j := 0; j < len(cur); j++ {
						if cur[j].y == -x {
							cur = append(cur[:j], cur[j+1:]...)
							break
						}
					}
				}
				a[x].b -= y
				if mn[a[x].a%k] <= int64(a[x].a) {
					cur = append(cur, pt{a[x].b, -x})
				}
			} else if len(cur) > 0 {
				fmt.Println(cur[0].x)
				used[-cur[0].y] = true
				cur = cur[1:]
			} else {
				fmt.Println(0)
			}
		}
	}

	for read() {
		solve()
	}
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Python/6110a86f4218962475ccd7a3475ce2eb.py
import random
import math
from queue import Queue
from collections import deque

INF = int(1e9)
INF64 = int(1e18)
MOD = INF + 7
EPS = 1e-9
PI = math.acos(-1.0)

N = 100 * 1000 + 13

n, m, k = 0, 0, 0
h = 0
a = []

def read():
    global h, n, m, k, a
    try:
        h, n, m, k = map(int, input().split())
        a = []
        for _ in range(n):
            x, y = map(int, input().split())
            a.append((x, y))
        return True
    except:
        return False

def add(x):
    global k, mn
    q = Queue()
    for i in range(k):
        q.put(i)
    while not q.empty():
        v = q.get()
        if mn[v] + x < mn[(v + x) % k]:
            mn[(v + x) % k] = mn[v] + x
            q.put((v + x) % k)

def recalc():
    global n, k, used, mn, a, cur
    cur.clear()
    for i in range(n):
        if not used[i] and mn[a[i][0] % k] <= a[i][0]:
            cur.append((a[i][1], -i))
    cur.sort(reverse=True)

def solve():
    global k, h, m, mn, used, a, cur
    mn = [h + 1] * k
    mn[1] = 1
    used = [False] * N
    cur = []
    recalc()
    for _ in range(m):
        t = int(input())
        if t == 1:
            x = int(input())
            add(x)
            recalc()
        elif t == 2:
            x, y = map(int, input().split())
            x -= 1
            if mn[a[x][0] % k] <= a[x][0]:
                cur.remove((a[x][1], -x))
            a[x] = (a[x][0], a[x][1] - y)
            if mn[a[x][0] % k] <= a[x][0]:
                cur.append((a[x][1], -x))
        elif len(cur) > 0:
            print(cur[0][0])
            used[-cur[0][1]] = True
            cur.pop(0)
        else:
            print(0)

while read():
    solve()
